core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/XMLReaderBatchSource.java

```diff
@@ -0,0 +1,308 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * XML Reader Batch Source Plugin
+ * It is used to read XML files from HDFS with specified file properties and filters.
+ * This reader emits XML event, specified by the node path property, for each file read.
+ */
+@Plugin(type = BatchSource.PLUGIN_TYPE)
+@Name("XMLReader")
+@Description("Batch source for XML read from HDFS")
+public class XMLReaderBatchSource extends ReferenceBatchSource<LongWritable, Object, StructuredRecord> {
+  private static final Logger LOG = LoggerFactory.getLogger(XMLReaderBatchSource.class);
+  private static final Gson GSON = new Gson();
+  private static final Type ARRAYLIST_PREPROCESSED_FILES  = new TypeToken<ArrayList<String>>() { }.getType();
+
+  public static final Schema DEFAULT_XML_SCHEMA = Schema.recordOf(
+    "xmlSchema",
+    Schema.Field.of("offset", Schema.of(Schema.Type.LONG)),
+    Schema.Field.of("filename", Schema.of(Schema.Type.STRING)),
+    Schema.Field.of("record", Schema.of(Schema.Type.STRING))
+  );
+
+  private final XMLReaderConfig config;
+
+  private KeyValueTable processedFileTrackingTable;
+  private FileSystem fileSystem;
+  private Path tempDirectoryPath;
+
+  public XMLReaderBatchSource(XMLReaderConfig config) {
+    super(config);
+    this.config = config;
+  }
+
+  @VisibleForTesting
+  XMLReaderConfig getConfig() {
+    return config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    super.configurePipeline(pipelineConfigurer);
+    config.validateConfig();
+    pipelineConfigurer.getStageConfigurer().setOutputSchema(DEFAULT_XML_SCHEMA);
+    pipelineConfigurer.createDataset(config.tableName, KeyValueTable.class.getName());
+  }
+
+  @Override
+  public void prepareRun(BatchSourceContext context) throws Exception {
+    Job job = JobUtils.createInstance();
+    Configuration conf = job.getConfiguration();
+    conf.set(XMLInputFormat.XML_INPUTFORMAT_PATH_NAME, config.path);
+    conf.set(XMLInputFormat.XML_INPUTFORMAT_NODE_PATH, config.nodePath);
+    if (StringUtils.isNotEmpty(config.pattern)) {
+      conf.set(XMLInputFormat.XML_INPUTFORMAT_PATTERN, config.pattern);
+    }
+    conf.set(XMLInputFormat.XML_INPUTFORMAT_FILE_ACTION, config.actionAfterProcess);
+    if (StringUtils.isNotEmpty(config.targetFolder)) {
+      conf.set(XMLInputFormat.XML_INPUTFORMAT_TARGET_FOLDER, config.targetFolder);
+    }
+
+    setFileTrackingInfo(context, conf);
+
+    //Create a temporary directory, in which XMLRecordReader will add file tracking information.
+    fileSystem = FileSystem.get(conf);
+    long startTime = context.getLogicalStartTime();
+    //Create temp file name using start time to make it unique.
+    String tempDirectory = config.tableName + startTime;
+    tempDirectoryPath = new Path(config.temporaryFolder, tempDirectory);
+    fileSystem.mkdirs(tempDirectoryPath);
+    fileSystem.deleteOnExit(tempDirectoryPath);
```

Comment:
This code create temporary folder on hdfs system, where XMLRecordReader will put the files containing name of processed XML file. These files in temporary folder will be read in onRunFinish() method. Folder get deleted after finishing the job, tested it on cluster CDAP 104.155.156.158.